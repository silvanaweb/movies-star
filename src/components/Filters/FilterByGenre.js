import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { setSelectedGenre } from "../../store/actions/filters";
import { CheckButton } from "../CheckButton/CheckButton";
import Select from "react-select";
import "./Filters.css";

const mapToSelectValue = value => ({
	value: value,
	label: value,
});

const FilterByGenre = ({ allgenres, setSelectedGenreFilter }) => {
	const [selectedGenres, setSelectedGenres] = useState([]);
	const [winWidth, setWinWidth] = useState(window.innerWidth);

	useEffect(() => {
		const handleResize = () => setWinWidth(window.innerWidth);
		window.addEventListener("resize", handleResize);
		return () => {
			window.removeEventListener("resize", handleResize);
		};
	}, []);

	useEffect(() => {
		const selectedGenresArr = Array.from(selectedGenres);
		// pass the genres to the filter state
		if (selectedGenresArr) {
			setSelectedGenreFilter(selectedGenresArr);
		}
	}, [selectedGenres]);

	const onSelectedGenre = ({ name, checked }) => {
		if (checked) {
			setSelectedGenres([...selectedGenres, name]);
		} else {
			setSelectedGenres(selectedGenres.filter(x => x !== name));
		}
		// after the state selectedGenres is updated,
		// the useEffect is called and there is where we call the action to set the filter  and then the filtered movies are updated
	};

	const onSelectedGenreMobile = selectedOptions => {
		setSelectedGenres(selectedOptions.map(o => o.name || o.value));
	};

	return (
		<div className="Filters__genre">
			<div className="content-container">
				{winWidth < 600 ? (
					<div className="Filters__genre__content__mobile">
						<Select
							name="filterGenres"
							placeholder="Filter by genres"
							value={selectedGenres.map(mapToSelectValue)}
							options={allgenres}
							onChange={onSelectedGenreMobile}
							isSearchable={false}
							isMulti
						/>
					</div>
				) : (
					<div className="Filters__genre__content">
						{allgenres.map((genre, idx) => (
							<div key={`genre.value${idx}`} className="Filters__genre__item">
								<CheckButton
									title={genre.name}
									name={genre.name}
									checked={selectedGenres.includes(genre.name)}
									onChange={onSelectedGenre}
								/>
							</div>
						))}
					</div>
				)}
			</div>
		</div>
	);
};

// map the state to props to get the genres list
const mapStateToProps = (state, props) => {
	return {
		allgenres: state.genres.map(g => ({ name: g.name, label: g.name })),
	};
};
// map the actions to props so that they can be called inside the component
const mapDispatchToProps = dispatch => ({
	setSelectedGenreFilter: genres => dispatch(setSelectedGenre(genres)),
});

export { FilterByGenre };

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(FilterByGenre);
