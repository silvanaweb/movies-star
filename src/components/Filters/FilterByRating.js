import React from "react";
import { connect } from "react-redux";
import { setSelectedRating } from "../../store/actions/filters";
import { RangeSlider } from "../RangeSlider/RangeSlider";

const FilterByRating = ({
  className,
  setSelectedRating,
  filters: { rating }
}) => {
  const onSelectedRatingChange = rating => {
    // dispatch to filters
    setSelectedRating(rating);
  };
  return (
    <div className={className}>
      <RangeSlider
        defaultValue={rating}
        min={1}
        max={10}
        step={0.5}
        onValueChange={onSelectedRatingChange}
      />
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    filters: state.filters
  };
};
const mapDispatchToProps = dispatch => ({
  setSelectedRating: rating => dispatch(setSelectedRating(rating))
});

export { FilterByRating };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FilterByRating);
