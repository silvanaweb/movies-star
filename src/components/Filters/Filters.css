.Filters {
  padding: var(--m-size) 0;
}

.Filters__content {
  display: grid;
  grid-template-columns: 6fr 6fr;
}

.Filters__genre {
  margin-bottom: var(--s-size);
}

@media (min-width: 600px) {
  .Filters {
    padding: var(--l-size) 0;
  }
  .Filters__content {
    grid-template-columns: 4fr 4fr 4fr;
  }
}

@media (min-width: 800px) {
  .Filters__content {
    grid-template-columns: 3fr 3fr 3fr 3fr;
  }
}
