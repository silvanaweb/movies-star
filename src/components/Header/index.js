import React from "react";
import { connect } from "react-redux";
import "./style.css";

const Header = ({ history }) => {
  return (
    <header className="Header">
      <div className="content-container">
        <div className="Header__content">header</div>
      </div>
    </header>
  );
};

const mapStateToProps = (state, props) => {
  return {};
};

export { Header, categories };

export default connect(mapStateToProps)(Header);
