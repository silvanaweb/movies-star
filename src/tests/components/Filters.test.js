import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import React from "react";
import { Filters } from "../../components/Filters";

let history, wrapper;

beforeEach(() => {
  history = { push: jest.fn() };
  wrapper = shallow(<Filters history={history} />);
});

test("should render Header correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});
