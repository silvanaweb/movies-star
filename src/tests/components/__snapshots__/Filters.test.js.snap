// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FilterByRating correctly 1`] = `
<div>
  <RangeSlider
    defaultValue={3}
    labels={
      Object {
        "tooltip": "rating from",
      }
    }
    max={10}
    min={0}
    onValueChange={[Function]}
    step={0.5}
  />
</div>
`;
